clang/test/CIR/Lowering/ThroughMLIR/cast.cir

```diff
@@ -0,0 +1,102 @@
+// RUN: cir-opt %s -cir-to-mlir | FileCheck %s -check-prefix=MLIR
+// RUN: cir-opt %s -cir-to-mlir -cir-mlir-to-llvm | mlir-translate -mlir-to-llvmir | FileCheck %s -check-prefix=LLVM
+
+!s32i = !cir.int<s, 32>
+!s16i = !cir.int<s, 16>
+!u32i = !cir.int<u, 32>
+!u16i = !cir.int<u, 16>
+!u8i  = !cir.int<u, 8>
+module {
+  cir.func @cast_int_to_bool(%i : !u32i) -> !cir.bool {
+    %1 = cir.cast(int_to_bool, %i : !u32i), !cir.bool
+    cir.return %1 : !cir.bool
+  }
+  cir.func @cast_integral_trunc(%i : !u32i) -> !u16i {
+    %1 = cir.cast(integral, %i : !u32i), !u16i
+    cir.return %1 : !u16i
+  }
+  cir.func @cast_integral_extu(%i : !u16i) -> !u32i {
+    %1 = cir.cast(integral, %i : !u16i), !u32i
+    cir.return %1 : !u32i
+  }
+  cir.func @cast_integral_exts(%i : !s16i) -> !s32i {
+    %1 = cir.cast(integral, %i : !s16i), !s32i
+    cir.return %1 : !s32i
+  }
+  cir.func @cast_integral_same_size(%i : !u32i) -> !s32i {
+    %1 = cir.cast(integral, %i : !u32i), !s32i
+    cir.return %1 : !s32i
+  }
+  cir.func @cast_floating_trunc(%d : !cir.double) -> !cir.float {
+    %1 = cir.cast(floating, %d : !cir.double), !cir.float
+    cir.return %1 : !cir.float
+  }
+  cir.func @cast_floating_extf(%f : !cir.float) -> !cir.double {
+    %1 = cir.cast(floating, %f : !cir.float), !cir.double
+    cir.return %1 : !cir.double
+  }
+  cir.func @cast_float_to_bool(%f : !cir.float) -> !cir.bool {
+    %1 = cir.cast(float_to_bool, %f : !cir.float), !cir.bool
+    cir.return %1 : !cir.bool
+  }
+  cir.func @cast_bool_to_int8(%b : !cir.bool) -> !u8i {
+    %1 = cir.cast(bool_to_int, %b : !cir.bool), !u8i
+    cir.return %1 : !u8i
+  }
+  cir.func @cast_bool_to_int(%b : !cir.bool) -> !u32i {
+    %1 = cir.cast(bool_to_int, %b : !cir.bool), !u32i
+    cir.return %1 : !u32i
+  }
+  cir.func @cast_bool_to_float(%b : !cir.bool) -> !cir.float {
+    %1 = cir.cast(bool_to_float, %b : !cir.bool), !cir.float
+    cir.return %1 : !cir.float
+  }
+  cir.func @cast_signed_int_to_float(%i : !s32i) -> !cir.float {
+    %1 = cir.cast(int_to_float, %i : !s32i), !cir.float
+    cir.return %1 : !cir.float
+  }
+  cir.func @cast_unsigned_int_to_float(%i : !u32i) -> !cir.float {
+    %1 = cir.cast(int_to_float, %i : !u32i), !cir.float
+    cir.return %1 : !cir.float
+  }
+  cir.func @cast_float_to_int_signed(%f : !cir.float) -> !s32i {
+    %1 = cir.cast(float_to_int, %f : !cir.float), !s32i
+    cir.return %1 : !s32i
+  }
+  cir.func @cast_float_to_int_unsigned(%f : !cir.float) -> !u32i {
+    %1 = cir.cast(float_to_int, %f : !cir.float), !u32i
+    cir.return %1 : !u32i
+  }
+}
+
+// MLIR: arith.cmpi ne, %arg0, %c0_i32
```

Comment:
Please check the name of the functions before checking the relevant content, example: `MLIR-LABEL: @cast_int_to_bool`. Also, for the content, SSA values need to be regexe'd, see testcases like `clang/test/CIR/Lowering/ptrstride.cir`.

Reply:
If I may add on top of that - it would be very helpful if you inline the check in the relevant piece of code. It makes each test more self contained and much easier to understand what you are trying to test. This makes the review process (and therefore - your self review as well! 😇 ) much more productive with higher quality.

For example:

```
  // MLIR-LABEL: @cast_int_to_bool
  cir.func @cast_int_to_bool(%i : !u32i) -> !cir.bool {
    // MLIR: arith.cmpi ne, %arg0, %c0_i32 (regexed, of course - I guess you also need to add a regex capture of the arith.const argument)

    %1 = cir.cast(int_to_bool, %i : !u32i), !cir.bool
    cir.return %1 : !cir.bool
  }
  ...
```

In general, I advise going through the MLIR testing guide  :) 
https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-tests

Reply:
> Please check the name of the functions before checking the relevant content, example: `MLIR-LABEL: @cast_int_to_bool`. Also, for the content, SSA values need to be regexe'd, see testcases like `clang/test/CIR/Lowering/ptrstride.cir`.

I will add function label check and regexe. Thanks for pointing me the test case for reference.

> If I may add on top of that - it would be very helpful if you inline the check in the relevant piece of code. It makes each test more self contained and much easier to understand what you are trying to test. This makes the review process (and therefore - your self review as well! 😇 ) much more productive with higher quality.
> 
> For example:
> 
> ```
>   // MLIR-LABEL: @cast_int_to_bool
>   cir.func @cast_int_to_bool(%i : !u32i) -> !cir.bool {
>     // MLIR: arith.cmpi ne, %arg0, %c0_i32 (regexed, of course - I guess you also need to add a regex capture of the arith.const argument)
> 
>     %1 = cir.cast(int_to_bool, %i : !u32i), !cir.bool
>     cir.return %1 : !cir.bool
>   }
>   ...
> ```
> 
> In general, I advise going through the MLIR testing guide :) https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-tests

It's a good point. I will inline the check lines to make it more self contained.
Thanks for providing an example and the link.